addons/l10n_ro_edi_stock/models/l10n_ro_edi_stock_etransport_interface.py

```diff
@@ -0,0 +1,640 @@
+
+    l10n_ro_edi_stock_state = fields.Selection(selection=DOCUMENT_STATES, compute='_compute_current_document_data')
+
+    l10n_ro_edi_stock_move_ids = fields.One2many(comodel_name='stock.move', compute='_compute_l10n_ro_edi_stock_move_ids', recursive=True)
+    l10n_ro_edi_stock_company_id = fields.Many2one(comodel_name='res.company', compute='_compute_l10n_ro_edi_stock_company_id', recursive=True)
+    l10n_ro_edi_stock_carrier_id = fields.Many2one(comodel_name='delivery.carrier', compute='_compute_l10n_ro_edi_stock_carrier_id', recursive=True)
+
+    l10n_ro_edi_stock_operation_type = fields.Selection(selection=OPERATION_TYPES, string="Operation Type")
+    l10n_ro_edi_stock_available_operation_scopes = fields.Char(compute='_compute_l10n_ro_edi_stock_available_operation_scopes')
+    l10n_ro_edi_stock_operation_scope = fields.Selection(selection=OPERATION_SCOPES, string="Operation Scope")
+
+    l10n_ro_edi_stock_vehicle_number = fields.Char(string="Vehicle Number", size=20)
+    l10n_ro_edi_stock_trailer_1_number = fields.Char(string="Trailer 1 Number", size=20)
+    l10n_ro_edi_stock_trailer_2_number = fields.Char(string="Trailer 2 Number", size=20)
+
+    l10n_ro_edi_stock_available_start_loc_types = fields.Char(compute='_compute_available_location_types')
+    l10n_ro_edi_stock_start_loc_type = fields.Selection(selection=LOCATION_TYPES, string="Location Type", default='location')
+
+    l10n_ro_edi_stock_available_end_loc_types = fields.Char(compute='_compute_available_location_types')
+    l10n_ro_edi_stock_end_loc_type = fields.Selection(selection=LOCATION_TYPES, string="Location Type", default='location')
+
+    # Data fields for every location type
+    l10n_ro_edi_stock_start_bcp = fields.Selection(selection=BORDER_CROSSING_POINTS, string="Border Crossing Point")
+    l10n_ro_edi_stock_start_customs_office = fields.Selection(selection=CUSTOMS_OFFICES, string="Custom Office")
+    l10n_ro_edi_stock_start_state_id = fields.Many2one(comodel_name='res.country.state', string="State", domain="[('country_id.code', '=', 'RO')]")
+    l10n_ro_edi_stock_start_city = fields.Char(string="City")
+    l10n_ro_edi_stock_start_street = fields.Char(string="Street")
+    l10n_ro_edi_stock_start_zip = fields.Char(string="Postal Code")
+    l10n_ro_edi_stock_start_other_info = fields.Char(string="Other Info")
+
+    l10n_ro_edi_stock_end_bcp = fields.Selection(selection=BORDER_CROSSING_POINTS, string="Border Crossing Point")
+    l10n_ro_edi_stock_end_customs_office = fields.Selection(selection=CUSTOMS_OFFICES, string="Customs Office")
+    l10n_ro_edi_stock_end_state_id = fields.Many2one(comodel_name='res.country.state', string="State", domain="[('country_id.code', '=', 'RO')]")
+    l10n_ro_edi_stock_end_city = fields.Char(string="City")
+    l10n_ro_edi_stock_end_street = fields.Char(string="Street")
+    l10n_ro_edi_stock_end_zip = fields.Char(string="Postal Code")
+    l10n_ro_edi_stock_end_other_info = fields.Char(string="Other Info")
```

Comment:
Idea: how about we create a new model: `l10n_ro.edi.stock.location`,

```py
class L10nRoEdiStockLocation(models.Model):
    _name = 'l10n_ro.edi.stock.location'
    
    bcp = fields.Selection(selection=BORDER_CROSSING_POINTS, string="Border Crossing Point")
    customs_office = fields.Selection(selection=CUSTOMS_OFFICES, string="Custom Office")
    state_id = fields.Many2one(comodel_name='res.country.state', string="State", domain="[('country_id.code', '=', 'RO')]")
    city = fields.Char(string="City")
    street = fields.Char(string="Street")
    zip = fields.Char(string="Postal Code")
    other_info = fields.Char(string="Other Info")
```

and here instead of having all these duplicate fields, you just initialize 2 field of that model:

```suggestion
    l10n_ro_edi_stock_start_location = fields.Many2one(comodel_name='l10n_ro.edi.stock.location')
    l10n_ro_edi_stock_end_location = fields.Many2one(comodel_name='l10n_ro.edi.stock.location')
```

and of course, adapt all the setattr and getattr. Maybe make a common helper function in the location model to reset, etc.

Reply:
I don't think that this is a good solution, it would of course get rid of the setattr and getattr calls but it would make the code base even more bloated. The problem lies in the fact that we need to show these fields on the picking (and batch) form views:

```xml
<field name="l10n_ro_edi_stock_start_bcp" readonly="l10n_ro_edi_stock_fields_readonly" invisible="l10n_ro_edi_stock_start_loc_type != 'bcp'" />
<field name="l10n_ro_edi_stock_start_customs_office" readonly="l10n_ro_edi_stock_fields_readonly" invisible="l10n_ro_edi_stock_start_loc_type != 'customs'"/>
<field name="l10n_ro_edi_stock_start_state_id" readonly="l10n_ro_edi_stock_fields_readonly" invisible="l10n_ro_edi_stock_start_loc_type != 'location'"/>
<field name="l10n_ro_edi_stock_start_zip" readonly="l10n_ro_edi_stock_fields_readonly" invisible="l10n_ro_edi_stock_start_loc_type != 'location'"/>
<field name="l10n_ro_edi_stock_start_city" string="City" readonly="l10n_ro_edi_stock_fields_readonly" invisible="l10n_ro_edi_stock_start_loc_type != 'location'"/>
<field name="l10n_ro_edi_stock_start_street" string="Street" readonly="l10n_ro_edi_stock_fields_readonly" invisible="l10n_ro_edi_stock_start_loc_type != 'location'"/>
<field name="l10n_ro_edi_stock_start_other_info" string="Other Info" readonly="l10n_ro_edi_stock_fields_readonly" invisible="l10n_ro_edi_stock_start_loc_type != 'location'"/>
```

If we were to extract the location model like you suggested, we would have to add related fields for every field of the location model to be able to use them in the picking/batch form views. IMO adding another model and more fields is not a good tradeoff for getting rid of setattr/getattr calls.

I also checked with Thomas (thbe) and he told me that from a security point of view these kind of setattr/getattr with prefixes is perfectly fine

Reply:
It's not just for getting rid of the setattr and getattr. It's because they look duplicated and it feels cleaner to refactor them to a model, since all of them share similar goal: for identifying the location.

For your problem, wouldn't this be possible?

```xml
<field name="l10n_ro_edi_stock_start_bcp" readonly="l10n_ro_edi_stock_fields_readonly" invisible="l10n_ro_edi_stock_start_location.loc_type != 'bcp'" />

Reply:
it would work if in the interface we added a related field;
```python
class L10nRoEdiStockInterface(models.Model):
    ...
    start_location_id = fields.Many2one(comodel='l10n_ro.edi.stock.location')
   l10n_ro_edi_stock_start_bcp = fields.Char(related='start_location_id.bcp')
```

and this would have to be done for every field in the location model